vars/buildDockerAndPublishImage.groovy

```diff
@@ -13,96 +13,73 @@ def call(String imageName, Map config=[:]) {
   // Customize Pod label to improve build analysis
   final String yamlPodDef = podYamlTemplate.replaceAll('\\$IMAGE_NAME', imageName).replaceAll('\\$?\\{IMAGE_NAME\\}', imageName)
 
-  // This function uses Declarative Syntax as we haven't met (yet) use case where only a subpart of the pipeline is needed: either we run it "whole" or not.
-  // It means that the function is acting more as a "template" of an initial declarative standalone pipeline, than a real life well-constructed library.
-  // A second reason is to lower the contribution bar by avoiding the "know very well Groovy" requirement (but this argument can be disputed as the test unit is pure Groovy...)
-  // If you feel like to rewrite this function in scripted, you can spend this effort of course :)
-  pipeline {
-    agent {
-      kubernetes {
-        inheritFrom 'jnlp-linux'
-        defaultContainer 'builder'
-        yaml yamlPodDef
-      } // kubernetes
-    } // agent
-
-    environment {
-      BUILD_DATE       = "${dockerConfig.buildDate}"
-      IMAGE_NAME       = "${dockerConfig.imageName}"
-      DOCKERFILE       = "${dockerConfig.dockerfile}"
-      PLATFORM         = "${dockerConfig.platform}"
-      SEMANTIC_RELEASE = "${dockerConfig.automaticSemanticVersioning}"
-    } // environment
-
-    stages {
-      stage('Next Version') {
-        when {
-          allOf {
-            environment name: 'SEMANTIC_RELEASE', value: 'true'
-            branch dockerConfig.mainBranch  
-          }
-        }
-        steps {
+  final Boolean semVerEnabled = dockerConfig.automaticSemanticVersioning && env.BRANCH_NAME == dockerConfig.mainBranch
+
+  podTemplate(
+    inheritFrom: 'jnlp-linux',
+    yaml: yamlPodDef,
+    defaultContainer: 'builder'
+  ) {
+    withEnv([
+      "BUILD_DATE=${dockerConfig.buildDate}",
+      "IMAGE_NAME=${dockerConfig.imageName}",
+      "DOCKERFILE=${dockerConfig.dockerfile}",
+      "PLATFORM=${dockerConfig.platform}",
+    ]) {
+      // Automatic tagging on principal branch is not enabled by default
+      if (semVerEnabled) {
+        stage('Next Version') {
           container('next-version') {
-            script {
-              nextVersion = sh(script:'jx-release-version', returnStdout: true).trim()
-              if (dockerConfig.metadataFromSh != '') {
-                metadata = sh(script: "${dockerConfig.metadataFromSh}", returnStdout: true).trim()
-                nextVersion = nextVersion + metadata
-              }
-            }
-          }
+            nextVersion = sh(script:'jx-release-version', returnStdout: true).trim()
+            if (dockerConfig.metadataFromSh != '') {
+              metadata = sh(script: "${dockerConfig.metadataFromSh}", returnStdout: true).trim()
+              nextVersion = nextVersion + metadata
+            } // if
```

Comment:
Assuming these variables are only being used in subsequent `sh` steps, better to avoid ever round-tripping through Groovy.
```groovy
sh 'jx-release-version > version.txt'
```
etc. And then of course if you do not have a specific reason for having intervening Pipeline steps, you can coalesce with subsequent `sh` steps.

This whole script still looks way too complicated. You could probably reduce it to a few lines of `podTemplate`, `withCredentials`, `sh`, and `recordIssues`, with all of the logic in some Bash script (or Ruby, Makefile, etc.) stored _elsewhere_.

Reply:
The goal to have separate stages for different "build step" is important for letting human understand what is going on when it fails. You're correct that we **might** be able to move a lot within the `Makefile` (for instance the conditional for the `test` goal) but whatever language it's written on, we still have to test it.

The pipeline unit help works out of the bo (except potential bugs) for this level of abstraction, while if we have to build a fully fledged framework, not really error-prone, for testing a Makefile based on its environment, it's possible but the effort is not really something that valuable. 

Thanks for the initial tip  though, we might be able to simplify some parts here (taking in account that it's currently being migrated into scripted syntax so the idioms have to be changed)